### PR TITLE
Allow the option use the call and create contract host functions

### LINKED ISSUE #29: CLI: Allow deploying and invoking a contract in CLI storage
### What problem does your feature solve?
There is no way to create a contract or call an existing contract without invoking a function in a WASM contract.

### What would you like to see?
A command stores a WASM contract to storage, and another command that allows invoking the stored contract through a contract ID.

### PR DESCRIPTION
**Pull request by sisuresh**
Resolves #29 and related to #22.

The two new commands take a JSON SCVec as an argument. I haven't updated how arguments are passed to the existing command that invokes WASM directly, but wanted to get input on this change before moving forward.

The user experience also isn't great, but I think we can make some improvements. The user needs to construct the JSON, which isn't really feasible by hand. For example, I tested this by using a rust test case and printed out the arguments that I wanted.

### CODE REVIEW

**Comment by leighmcculloch** [comment]
cc @paulbellamy who is working on:
- #22.

File: README.md
Comment (leighmcculloch) [inline README.md:34]: What does `--file <JSONARGS>` mean?

File: README.md
Comment (leighmcculloch) [inline README.md:16]: It's not obvious that `vm-fn` means a contract fn for a user uploaded contract. I think we should keep the top-level subcommand `invoke` as to mean invoking a contract fn. We should try to abstract some of the oddities of the fact that all fns are host called, the CLI doesn't need to expose this.

File: README.md
Comment (leighmcculloch) [inline README.md:34]: What does `--file <JSONARGS>` mean? It looks like it is JSON instead of a WASM file. We should avoid overloading a single argument as the behavior will be confusing to anyone who hasn't in detail read the docs for the command.

Comment (leighmcculloch) [inline README.md:34]: I think we should replace `invoke create-contract` with a `deploy`. We plan to add functionality for deploying to a running network (#41), and this function is identical to that except that we're not submitting a tx and it is all happening locally. As part of #41 we can figure out how to distinguish them.

File: README.md
Comment (leighmcculloch) [inline README.md:43]: Similar to comment to above, I think we should avoid overloading `--file` with anything other than a WASM file. This is ideally `cli invoke --file <WASM> --fn <...>` when a file is being passed as a short hand for deploy and execute, and then in the manual deploy case `cli invoke --contract-id <...> --fn <...>`. i.e. `--file` and `--contract-id` are mutually exclusive.

**Review by leighmcculloch — COMMENTED** [review]
I left a few comments inline. The capabilities look good, but I think we need to present this in a manner that abstracts the idiosyncrasies of `InvokeHostFunctionOp` rather than requires the developer to interact with them.

File: README.md
Comment (sisuresh) [reply README.md]: I removed the subcommands I added.

File: README.md
Comment (sisuresh) [reply README.md]: All commands now take JSON strings as args.

Comment (sisuresh) [reply README.md]: Done. `deploy` just commits a contractID and WASM blob to storage. This can easily be extended to  create a contract using the host function in the future.

**Comment by sisuresh** [comment]
@leighmcculloch The github checks are failing because I stopped using `strval`. I'll fix that if we agree on the change in this PR.

File: README.md
Comment (leighmcculloch) [inline README.md:18]: Is the `vm-fn` still part of the invoke command?

File: README.md
Comment (leighmcculloch) [inline README.md:24]: If it is low effort to revert the strval change and move it to a separate PR, I think we should address that change in a separate PR so that the majority of the changes in this PR are not blocked by #22. This format is close, but not exactly what I think @paulbellamy was proposing, as he was proposing `--arg 32`.

**Review by leighmcculloch — COMMENTED** [review]
If you can break out the strval change to a separate PR, I think we can merge this. Or maybe we just merge it as this is, and @paulbellamy can iterate from here on improving the format, as it probably doesn't matter if he works from here or from strval.

File: README.md
Comment (leighmcculloch) [reply README.md:24]: Spoke to @paulbellamy who said he can work from here with #22, so let's merge this as is.

File: README.md
Comment (sisuresh) [reply README.md]: Nope. Removed.

File: README.md
Comment (leighmcculloch) [inline README.md:49]: This format reinforces the need for us to do stellar/xdrgen#125, stellar/xdrgen#118, stellar/xdrgen#119.

File: src/invoke.rs
Comment (leighmcculloch) [inline src/invoke.rs:72]: I think this should be resolved by always requiring `--contract-id`, and by essentially having the file uploaded overwrite whatever is already stored at that ID. This type of behavior will have the CLI behave a little more intuitively when someone executes a contract by file, then executes a second contract by file that references the former. In this case the former is also stored and can be executed by the second contract.

File: README.md
Comment (leighmcculloch) [inline README.md:37]: Is there a way we could support smaller IDs? One thing I appreciated using some other CLI tools in this space was when they would allow me to use invalid values for things like a contract hash, like `0x1`, `0x2`, etc which is helpful in testing. I think within the CLI tool we could simply right-fill with zeros to the 32 byte size we need.

**Review by leighmcculloch — APPROVED** [review]
Nice, this looks very intuitive. I left a few comments, but to avoid this PR getting stale I think we could merge this and follow up with them if you agree they're worth doing.

File: src/invoke.rs
Comment (sisuresh) [reply src/invoke.rs:72]: I'm not sure I understand. This block is only executed if we're executing a WASM blob directly, so the contract won't be stored. The use case of specifying the contractID here is only useful if the contract uses the contractID internally. I don't think we should write the contract to storage using the `invoke` command. That's what `deploy` is for.

File: README.md
Comment (sisuresh) [reply README.md:37]: Yeah I think this is a great idea. Opened an issue here https://github.com/stellar/stellar-contract-cli/issues/45.

File: src/invoke.rs
Comment (leighmcculloch) [reply src/invoke.rs:72]: The invocation will still be operating on the ledger state, so I think it should be possible for a contract to access state of the given contract ID from prior runs. The fact we're disallowing an ID to be provided just means it is always accessing the 0x0 state.

>I don't think we should write the contract to storage using the invoke command. That's what deploy is for.
Having to run deploy manually is inconvenient for local testing. I think the `invoke` command when provided with the file should be a shortcut for deploying that file, to the contract ID, and invoking it. This is essentially what it is doing with the zero ID, but without persisting.

This will make for convenient workflows where a developer can run a command sequence like the below, and where the invocations of the my.wasm and my2.wasm can be associated with distinct contract data that doesn't overlap. It will then also allow my2.wasm to call my.wasm if that's useful.

```
stellar-contract-cli invoke --id 0x1 --file my.wasm --fn go --arg 1
stellar-contract-cli invoke --id 0x1 --file my.wasm --fn go --arg 2
stellar-contract-cli invoke --id 0x2 --file my2.wasm --fn doit
stellar-contract-cli invoke --id 0x1 --file my.wasm --fn go --arg 3
```

The alternative, requiring explicit deploy commands, is less convenient, especially when iterating on a contract, where you might deploy after each run:

```
stellar-contract-cli deploy --id 0x1 --file my.wasm
stellar-contract-cli invoke --id 0x1 --fn go --arg 1
stellar-contract-cli deploy --id 0x1 --file my.wasm
stellar-contract-cli invoke --id 0x1 --fn go --arg 2
stellar-contract-cli deploy --id 0x2 --file my2.wasm
stellar-contract-cli invoke --id 0x2 --fn doit
stellar-contract-cli deploy --id 0x1 --file my.wasm
stellar-contract-cli invoke --id 0x1 --fn go --arg 3
```

Comment (sisuresh) [reply src/invoke.rs:72]: Ah I see what you're saying. Remove `deploy`, always specify `contractID` so you can access any contract data, and write the contract specified in `invoke` to storage. Yeah I think this is a good idea. I'll make this change now.

Comment (sisuresh) [reply src/invoke.rs]: Done. I kept `deploy` as we discussed, but `invoke` now deploys the contract as well if file is specified.

File: README.md
Comment (2opremio) [reply README.md:36]: https://github.com/stellar/soroban-cli/pull/49

**Pull request merged**
